pygfx/geometries/_text.py

```diff
@@ -0,0 +1,167 @@
+it produces glyph indices and positions. There may be less glyphs than
+Unicode characters because of ligatures etc. It also makes sure to add
+the glyph to the global glyph atlas.
+
+The TextGeometry takes the text parts and combines them into a single
+string of glyphs. It is responsible for the positioning.
+
+Finally, the renderer will render the glyphs on screen.
+
+"""
+
+
+def text_geometry(
+    text=None, markdown=None, max_width=None, line_height=None, text_align=None
+):
+    """Generate text geometry."""
+
+    # === Itemization - generate a list of TextParts objects
+    parts = []
+    if text:
+        if not isinstance(text, str):
+            raise TypeError("Text must be a Unicode string.")
+        parts.append(TextPart(text))
+    if markdown:
+        raise NotImplementedError()
+
+    # === Reordering - put parts in the right order
+    pass  # we're assuming LTR for now
+
+    return TextGeometry(
+        parts, max_width=max_width, line_height=line_height, text_align=text_align
+    )
+
+
+class TextPart:
+    """A text part represents a piece of text that is formatted in a uniform
+    way. One piece of text can consists of multiple parts. This class
+    is responsible for shaping the text.
+    """
+
+    def __init__(self, text, font=None, font_size=12, weight=400, italic=False):
+
+        # === Shaping - generate indices and positions
+
+        self.indices = np.zeros((len(text),), np.uint32)
+        self.positions = np.zeros((len(text), 3), np.float32)
+
+        for i in range(len(text)):
+            self.positions[i] = i * font_size, 0, 0
+
+
+class TextGeometry(Geometry):
+    def __init__(self, text_parts, max_width=0, line_height=1.2, text_align="left"):
+        super().__init__()
+
+        # Check incoming parts
+        self._text_parts = tuple(text_parts)
+        for part in self._text_parts:
+            if not isinstance(part, TextPart):
+                raise TypeError("TextGeometry only accepts TextPart objects.")
+
+        # Set props
+        self.max_width = max_width
+        self.line_height = line_height
+        self.text_align = text_align
+
+        # Compose the parts in a single geometry
+        indices_arrays = []
+        positions_arrays = []
+        glyph_count = 0
+        for part in self._text_parts:
+            assert part.indices.dtype == np.uint32
+            assert part.positions.dtype == np.float32
+            assert part.positions.shape == (part.indices.size, 3)
+            part.offset = glyph_count
+            glyph_count += part.indices.size
+            indices_arrays.append(part.indices)
+            positions_arrays.append(part.positions)
+
+        # Store
+        self.indices = Buffer(np.concatenate(indices_arrays, 0))
+        self.positions = Buffer(np.concatenate(positions_arrays, 0))
+
+    @property
+    def max_width(self):
+        """The maximum width of the text (in model coords). Text will
```

Comment:
I'm not sure what model space means... :) Font space?

Reply:
Good question. The docstring is wrong - it would depend on the `screen_space` prop of the material. This is worth some more thought and discussion ...

Reply:
I would expect max width to be expressed in some font relative unit, so that line wrapping is not going to have different results based on how the text is rendered (billboard-style or 3d).

Reply:
Yeah exactly. It operates in the same space as the font size and other positions.